File: bot_comm/src/ssltest.py
import sys
sys.path.append('/opt/ros/jade/lib/python2.7/dist-packages/')
import rospy
import serial, os
from pylab import *
import math
import time 
sys.path.append('../../')
frame = 1
ti = time.time()            
# # sys.path.append('/home/aif/ROBOCUP_SSL_WS/devel/lib/python2.7/dist-packages/krssg_ssl_msgs/msg/')
# from krssg_ssl_msgs.msg import gr_Commands
ser = None

# # ser = serial.Serial('/dev/ttyUSB1',bytesize=32)
# ser = serial.Serial('/dev/ttyUSB1')
file = ''
try :
    try:
        # global ser, file
        ser = serial.Serial('/dev/ttyUSB0',115200)
        file = '/dev/ttyUSB0'
    except:
        # global ser
        ser = serial.Serial('/dev/ttyUSB1',115200)
        file = '/dev/ttyUSB1'
    # ser = serial.Serial('/dev/ttyUSB1',bytesize=32)
    print(ser)

except :
    print("Couldn't open the port\n")


Team_ID = 127  # Team Blue
theta = [30.0,150.0,225.0,315.0] # Bot angles
bot_radius = 0.087
bot_wheel_radius = 0.025
max_vel_wheel = 5000.0 # in rpm
# global buf
buf = [None]*32
for i in range(32):
    buf[i] = 0
buf[0] = Team_ID
FACTOR_T = 40
FACTOR_N = 40
FACTOR_W = 90
v_4_wheel = [None, None, None, None]
ch_buff = ''
from os import stat


def vel_convert(vel_3_wheel):
    vx = vel_3_wheel[0]
    vy = vel_3_wheel[1]
    vw = -vel_3_wheel[2]

    print("vx",vx,"vy",vy,"vw",vw)
    for i in range(4):
        v_4_wheel[i] = ((bot_radius*vw) - (vx*math.sin(theta[i]*math.pi/180.0)) + (vy*math.cos(theta[i]*math.pi/180.0)))/(bot_wheel_radius * math.pi)
    for i in range(4):
        if v_4_wheel[i] > 0 :
            v_4_wheel[i] = 126 + ((v_4_wheel[i]-max_vel_wheel)*126) / max_vel_wheel
        else :
            v_4_wheel[i] = 255 + (v_4_wheel[i]*129) / max_vel_wheel
    return v_4_wheel

def gr_Commands_CB(bot_id, velnormal, veltangent, velangular, spinner=False,kickspeedx=0):
    global buf,ser
    buff = ''
    # for i in xrange(1,32):
    #     buf[i] = 20
    # for i in xrange(4):
    #     buf[i+start] = vel_convert([veltangent,velnormal,velangular])
    # buf[11],   buf[12],  buf[13],  buf[14] = vel_convert([veltangent,velnormal,velangular])

    v_4_wheel = vel_convert([veltangent,velnormal,velangular])
    start = 1 + bot_id*5
    for i in xrange(4):
        buf[i+start] = int(v_4_wheel[i])
    # buf[1],   buf[2],  buf[3],  buf[4] = vel_convert([0,-50,0])
    for i in xrange(11,15):
        buf[i] = int(buf[i])

    for i in xrange(1,5):
        buf[i] = int(buf[i])
    for i in xrange(32):
        buff += chr(int(buf[i])%256)

    print(buf)
    ser.write(buff)

if __name__ == '__main__':
    import argparse

    parser = argparse.ArgumentParser(add_help=True)

    parser.add_argument('--id','-i', default=0)
    parser.add_argument('--vx','-x', default=0)
    parser.add_argument('--vy','-y', default=-40)
    parser.add_argument('--vw','-w', default=0)
    parser.add_argument('--range','-r',type=str,default = "none")
    
    args = parser.parse_args()

    while True:
        if "none" in args.range:
            gr_Commands_CB(bot_id=args.id, velnormal=args.vy, veltangent=args.vx, velangular=args.vw)
        else:
            bots = map(int,args.range.split('-'))
            for i in range(bots[0],bots[1]):
                gr_Commands_CB(bot_id=i, velnormal=args.vy, veltangent=args.vx, velangular=args.vw)

File: bot_comm/src/ssltest.cpp
#include <stdio.h>
#include "serial.h"
#include <stdlib.h>
#include <iostream>
#include <cmath>
#include <unistd.h>
#include <fstream>

#define RATIO 40    
#define FACTOR_T RATIO
#define FACTOR_N RATIO
#define FACTOR_W 90

/*
    0 = 00
    1 = 01
    2 = 10
    3 = 11
    <DRIBBLER><KICKER>
*/
#define _DRIBBLER   1<<1
#define _KICKER     1<<0

using namespace std;

const int TEAM_ID = 127;  //blue
const double Radius=0.087; //bot radius
const double radius=0.025; //wheel radius
const double PI=3.141592653589793;
double max_vel_wheel=5000.0; //this is in rpm
double theta[4]={30.0,150.0,225.0,315.0};

HAL::Serial serial;

void vel_convert(const double* vel_xyw, double* vel_wheel)
{
    double vx=vel_xyw[0];
    double vy=vel_xyw[1];
    double vw=vel_xyw[2];
    vw*=-1;
    int i;
    vel_wheel[0] =   (( (Radius * vw) - (vx * sin(theta[0])) + (vy * cos(theta[0]))) )/radius;
    vel_wheel[1] =   (( (Radius * vw) - (vx * sin(theta[1])) + (vy * cos(theta[1]))) )/radius;
    vel_wheel[2] =   (( (Radius * vw) - (vx * sin(theta[2])) + (vy * cos(theta[2]))) )/radius;
    vel_wheel[3] =   (( (Radius * vw) - (vx * sin(theta[3])) + (vy * cos(theta[3]))) )/radius;

    for (i = 0; i < 4; i++)
        vel_wheel[i]=vel_wheel[i]/PI;
    
    for (i = 0; i < 4; i++){        
        if(vel_wheel[i]>0)
            vel_wheel[i] = 126 + ((vel_wheel[i]-max_vel_wheel)*(126.0))/max_vel_wheel;
        else
            vel_wheel[i] = 256 - ((vel_wheel[i]-0)*(129.0-256.0))/max_vel_wheel;
    }
}

void CallBack(int ID, double* vel_wheel)
{
    unsigned static char buf[32];
    buf[0] = TEAM_ID;
    int _start = 1+ID*5, c=0;
    for(int i=_start;i<_start + 4;i++, c++)
        buf[i] = (int)vel_wheel[c];
    buf[_start + 4] = 0;
    for(int i=0;i<32;i++){
        cout<<(int)buf[i]<<" ";
    }
    cout<<endl;
    // if()
    cout<<serial.Write(buf,32)<<endl;
}

int main(int argc, char *argv[])
{
    //Converting angles to radians
    for (int i = 0; i < 4; ++i)
        theta[i]=theta[i]*PI/180;

    if(!serial.Open("/dev/ttyUSB0", 115200)) {
        printf("Could not open the fucking parts.\n");
        exit(0);
    }
    

    double vel_wheel[4];
   /* for(int i=0;i<4;i++)
        cin>>vel_wheel[i];
    while(1){
        CallBack(0, vel_wheel);
    }*/
    int c = 0;
    // double d1[4] = {0,0,0,0}, d2[4] = {20, 20, 20, 20}, d3[4] = {229, 0, 27, 0};
    double d0[4] = {0, 0, 0, 0};
    double d1[4] = {80, 0, 0, 0};
    double d2[4] = {0, 80, 0, 0};
    double d3[4] = {0, 0, 80, 0};
    double d4[4] = {0, 0, 0, 80};

//while(1){
/*
    c=0;
    CallBack(1, d4);
    /*while(c<400000){
        c++;
    }
cout<<"change 1"<<endl;
    usleep(4000000);
    c = 0;
    CallBack(1, d1);
    // while(c<40000){
    //      CallBack(1, d1);
    //     c++;
    // }
cout<<"change 2"<<endl;
     usleep(4000000);
    c = 0;
        CallBack(1, d2);
    usleep(4000000);
    // while(c < 40000) {
    // CallBack(1, d2);
    //     c++;
    // }
cout<<"change 3"<<endl;
    c=0;
        CallBack(1, d3);
     usleep(4000000);
        // while(c < 40000){
        //     CallBack(1, d3);
        //     c++;
        // }
cout<<"change 4"<<endl;
}
*/
    double v3d[3] = {0, 0, 0},drib;
    double wheel4d[4];
    using namespace std;
    int x;
    std::cin>>v3d[0]>>v3d[1]>>v3d[2]>>drib>>x;

     vel_convert(v3d, wheel4d);
    
    unsigned static char buf[32];
    buf[0] = TEAM_ID;

    for(int i=1;i<5;i++)
        buf[i] = (int)wheel4d[i-1];

    buf[5+x*5] = drib;


    cout<<endl;
    for(int i=0;i<32;i++){
        cout<<(int)buf[i]<<" ";
    }

    while(1) {
        serial.Write(buf,32);
        for(int i=0;i<32;i++){
            cout<<(int)buf[i]<<" ";
        }
        cout<<endl;
    }
    return 0;
}
